src/main/java/chatbot/Duke.java

```diff
@@ -0,0 +1,176 @@
+package chatbot;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+
+/**
+ * Thrown to indicate that the description for a Todo task cannot be empty.
+ */
+class EmptyTodoException extends Exception {
+
+  public EmptyTodoException() {
+    super("Todo cannot have empty description .");
+  }
+}
+
+/**
+ * Thrown to indicate that the description for a Deadline task cannot be empty.
+ */
+class EmptyDeadlineException extends Exception {
+
+  public EmptyDeadlineException() {
+    super("Deadline cannot have empty description.");
+  }
+}
+
+/**
+ * Thrown to indicate that the description for an Event task cannot be empty.
+ */
+class EmptyEventException extends Exception {
+
+  public EmptyEventException() {
+    super("Event cannot have empty description.");
+  }
+}
+
+/**
+ * Thrown to indicate that the user input is invalid.
+ */
+class InvalidException extends Exception {
+
+  public InvalidException() {
+    super("Sorry, invalid input.");
+  }
+}
+
+/**
+ * Represents a generic task with a description and a completion status.
+ */
+class Task {
+
+  private String description;
+  private boolean isDone;
+
+  public Task(String description) {
+    this.description = description;
+    this.isDone = false;
+  }
+
+  public String getStatusIcon() {
+    return (isDone ? "[X] " : "[ ] "); // mark done task with X
+  }
+
+  public void markAsDone() {
+    this.isDone = true;
+  }
+
+  public void markAsUndone() {
+    this.isDone = false;
+  }
+
+  @Override
+  public String toString() {
+    return (this.getStatusIcon() + this.description);
+  }
+
+  public String getDescription() {
+    return description;
+  }
+}
+
+/**
+ * Represents a Todo task.
+ */
+class Todo extends Task {
+
+  public Todo(String description) {
+    super(description);
+  }
+
+  @Override
+  public String toString() {
+    return ("[T]" + super.toString());
+  }
+}
+
+/**
+ * Represents a Deadline task with a due date.
+ */
+class Deadline extends Task {
+
+  protected String by;
```

Comment:
Maybe a more specific variable name (like deadline?), I'm not too sure what 'by' refers to here immediately (although I do get it means 'by [deadline]'.